### PR TITLE
Warn if /var disk space is full and add a solution message

### LINKED ISSUE #9024: Warn when /var disk space is limited
To replicate:

1. Set Docker to 8GB
2. Gather minikube dependencies: `minikube start --download-only`
3. Fill the Docker image cache  up until `docker system df` is full. If you need help doing so, I found this works: 

```
seq 32 1 | xargs -n1 -I{} docker pull fedora:{}
seq 8 1 | xargs -n1 -I{} docker pull oraclelinux:{} 
seq 20 1 | xargs -n1 -I{} echo docker pull ubuntu:{}:10
seq 20 1 | xargs -n1 -I{} echo docker pull ubuntu:{}:04
```

You'll probably see at this point errors, such as `no space left on device` -- that's a good thing!

4. Run `minikube start`

## Here's what I see 

Notice that there is no warning:

```
😄  minikube v1.12.2 on Darwin 10.15.5
✨  Automatically selected the docker driver. Other choices: hyperkit, vmware, parallels, virtualbox, vmwarefusion
👍  Starting control plane node minikube in cluster minikube
🔥  Creating docker container (CPUs=2, Memory=3892MB) ...
🐳  Preparing Kubernetes v1.18.3 on Docker 19.03.8 ...
🔎  Verifying Kubernetes components...
🌟  Enabled addons: default-storageclass, storage-provisioner
🏄  Done! kubectl is now configured to use "minikube"
```

/var is completely full!

```
% minikube ssh "df -h"
Filesystem      Size  Used Avail Use% Mounted on
overlay         7.9G  7.8G     0 100% /
tmpfs            64M     0   64M   0% /dev
tmpfs           2.0G     0  2.0G   0% /sys/fs/cgroup
shm              64M     0   64M   0% /dev/shm
/dev/vda1       7.9G  7.8G     0 100% /var
tmpfs           2.0G  8.5M  2.0G   1% /run
tmpfs           2.0G   12K  2.0G   1% /tmp
/dev/root       390M  390M     0 100% /usr/lib/modules
tmpfs           5.0M     0  5.0M   0% /run/lock
```

node shows as ready:

```
NAME       STATUS   ROLES    AGE     VERSION
minikube   Ready    master   3m21s   v1.18.3
```

Describe shows no disk pressure:

```
Name:               minikube
Roles:              master
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/os=linux
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=minikube
                    kubernetes.io/os=linux
                    minikube.k8s.io/commit=be7c19d391302656d27f1f213657d925c4e1cfc2-dirty
                    minikube.k8s.io/name=minikube
                    minikube.k8s.io/updated_at=2020_08_18T12_26_43_0700
                    minikube.k8s.io/version=v1.12.2
                    node-role.kubernetes.io/master=
Annotations:        kubeadm.alpha.kubernetes.io/cri-socket: /var/run/dockershim.sock
                    node.alpha.kubernetes.io/ttl: 0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Tue, 18 Aug 2020 12:26:40 -0700
Taints:             <none>
Unschedulable:      false
Lease:
  HolderIdentity:  minikube
  AcquireTime:     <unset>
  RenewTime:       Tue, 18 Aug 2020 12:29:54 -0700
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  MemoryPressure   False   Tue, 18 Aug 2020 12:26:55 -0700   Tue, 18 Aug 2020 12:26:36 -0700   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Tue, 18 Aug 2020 12:26:55 -0700   Tue, 18 Aug 2020 12:26:36 -0700   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Tue, 18 Aug 2020 12:26:55 -0700   Tue, 18 Aug 2020 12:26:36 -0700   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Tue, 18 Aug 2020 12:26:55 -0700   Tue, 18 Aug 2020 12:26:55 -0700   KubeletReady                 kubelet is posting ready status
Addresses:
  InternalIP:  172.17.0.3
  Hostname:    minikube
Capacity:
  cpu:                2
  ephemeral-storage:  8190392Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             4035432Ki
  pods:               110
Allocatable:
  cpu:                2
  ephemeral-storage:  8190392Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             4035432Ki
  pods:               110
System Info:
  Machine ID:                 b95ed2bfb4c74ec9b5b19a9e71395509
  System UUID:                5e5d6d07-9c65-40ac-90e3-2e5e74257084
  Boot ID:                    2a9ce373-3ed6-4829-acc5-3a49f09ddb6f
  Kernel Version:             4.19.76-linuxkit
  OS Image:                   Ubuntu 20.04 LTS
  Operating System:           linux
  Architecture:               amd64
  Container Runtime Version:  docker://19.3.8
  Kubelet Version:            v1.18.3
  Kube-Proxy Version:         v1.18.3
Non-terminated Pods:          (7 in total)
  Namespace                   Name                                CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                   ----                                ------------  ----------  ---------------  -------------  ---
  kube-system                 coredns-66bff467f8-4vm7b            100m (5%)     0 (0%)      70Mi (1%)        170Mi (4%)     3m15s
  kube-system                 etcd-minikube                       0 (0%)        0 (0%)      0 (0%)           0 (0%)         3m18s
  kube-system                 kube-apiserver-minikube             250m (12%)    0 (0%)      0 (0%)           0 (0%)         3m18s
  kube-system                 kube-controller-manager-minikube    200m (10%)    0 (0%)      0 (0%)           0 (0%)         3m18s
  kube-system                 kube-proxy-k6lnp                    0 (0%)        0 (0%)      0 (0%)           0 (0%)         3m15s
  kube-system                 kube-scheduler-minikube             100m (5%)     0 (0%)      0 (0%)           0 (0%)         3m18s
  kube-system                 storage-provisioner                 0 (0%)        0 (0%)      0 (0%)           0 (0%)         3m18s
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests    Limits
  --------           --------    ------
  cpu                650m (32%)  0 (0%)
  memory             70Mi (1%)   170Mi (4%)
  ephemeral-storage  0 (0%)      0 (0%)
  hugepages-1Gi      0 (0%)      0 (0%)
  hugepages-2Mi      0 (0%)      0 (0%)
Events:
  Type     Reason                   Age    From                  Message
  ----     ------                   ----   ----                  -------
  Normal   Starting                 3m20s  kubelet, minikube     Starting kubelet.
  Normal   NodeHasSufficientMemory  3m20s  kubelet, minikube     Node minikube status is now: NodeHasSufficientMemory
  Normal   NodeHasNoDiskPressure    3m20s  kubelet, minikube     Node minikube status is now: NodeHasNoDiskPressure
  Normal   NodeHasSufficientPID     3m20s  kubelet, minikube     Node minikube status is now: NodeHasSufficientPID
  Normal   NodeAllocatableEnforced  3m19s  kubelet, minikube     Updated Node Allocatable limit across pods
  Warning  readOnlySysFS            3m14s  kube-proxy, minikube  CRI error: /sys is read-only: cannot modify conntrack limits, problems may arise later (If running Docker, see docker issue #24000)
  Normal   Starting                 3m14s  kube-proxy, minikube  Starting kube-proxy.
  Normal   NodeReady                3m9s   kubelet, minikube     Node minikube status is now: NodeReady
```

It seems that the kubelet is not configured to track /var.

### PR DESCRIPTION
**Pull request by priyawadhwa**
closes #9024

Current output on my machine for failure looks like this:

```
$ minikube start --driver docker
$ minikube start --driver docker
😄  minikube v1.12.3 on Darwin 10.15.6
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4893MB) ...

❌  [NO_SPACE_ON_DEVICE]  docker daemon out of memory. No space left on device
💡  Suggestion: Try at least one of the following to free up space on the device:

	1. Run "docker system prune"
	2. Increase the amount of memory allocated to Docker for Desktop via 
		Docker icon > Preferences > Resources > Disk Image Size
	3. Run "minikube ssh -- docker system prune" if using the docker container runtime

⁉️   Related issue: https://github.com/kubernetes/minikube/issues/9024
```

### CODE REVIEW

File: pkg/minikube/problem/err_map.go [hunk 1]
```diff
@@ -578,3 +578,12 @@ var stateProblems = map[string]match{
 		Issues: []int{7256},
 	},
 }
+
+// dockerProblems are issues relating to issues with the docker driver
+var dockerProblems = map[string]match{
+	"NO_SPACE_ON_DEVICE": {
+		Regexp: re(`No space left on device`),
+		Advice: "Run 'docker system prune' to delete unused data and free up space on the device",
```
Comment (medyagh) [inline pkg/minikube/problem/err_map.go:586]: I suggest if it is a docker desktop it should tell the user to increase their docker desktop disk size
We have func to check if driver is docker desktop


Setting, resources ... Disk ....

File: pkg/minikube/machine/start.go [hunk 2]
```diff
@@ -202,6 +205,26 @@ func timedCreateHost(h *host.Host, api libmachine.API, t time.Duration) error {
...
+	output, err := h.RunSSHCommand("df -h /var | awk 'NR==2{print $5}'")
+	if err != nil {
+		glog.Warning("error running df -h /var: %v", err)
+	}
+	output = strings.Trim(output, "\n")
+	percentageFull, err := strconv.Atoi(output[:len(output)-1])
+	if err != nil {
+		glog.Warning("error getting % of /var that is free: %v", err)
+	}
+	if percentageFull >= 99 {
```
Comment (medyagh) [inline pkg/minikube/machine/start.go:223]: I think we should also warm them at 80% but maybe with more friendly they hey you are getting close to fill up the disk.

Comment (priyawadhwa) [reply pkg/minikube/machine/start.go:224]: Cool, added that in

File: pkg/minikube/problem/err_map.go (hunk 1 above)
Comment (priyawadhwa) [reply pkg/minikube/problem/err_map.go]: Added that in

File: pkg/minikube/machine/start.go [hunk 3]
```diff
@@ -202,6 +206,30 @@ func timedCreateHost(h *host.Host, api libmachine.API, t time.Duration) error {
...
+	percentageFull, err := strconv.Atoi(output[:len(output)-1])
+	if err != nil {
+		glog.Warningf("error getting percentage of /var that is free: %v", err)
+	}
+	if percentageFull >= 99 {
+		exit.WithError("", fmt.Errorf("docker daemon out of memory. No space left on device"))
+	}
+
+	if percentageFull > 80 {
+		out.WarningT("The docker daemon is almost out of memory, run 'docker system prune' to free up space")
```
Comment (medyagh) [inline pkg/minikube/machine/start.go:229]: The advice we give to the user should be based the following

// for windows and mac docker driver users
if  driver.IsDockerDesktop(drvName) // 

Docker Desktop >. Preferences > Resources  > Increase Disk Image


if they are VM driver and they are using Docker Runtime
then they should do
- minikube ssh docker system prune
- upgrade their system's disks size or resize partitions

Or if they are VM driver and Not Docker Runtime they should or None Driver
- upgrade their system's disks size or resize partitions

Comment (priyawadhwa) [reply pkg/minikube/machine/start.go:229]: Right now I'm only handling the case for the docker driver (first couple lines of this check)

Comment (priyawadhwa) [reply pkg/minikube/machine/start.go:229]: Added the minikube ssh tip in, the output should be in the PR description

File: pkg/minikube/machine/start.go [hunk 4]
```diff
@@ -202,6 +206,30 @@ func timedCreateHost(h *host.Host, api libmachine.API, t time.Duration) error {
...
+	output = strings.Trim(output, "\n")
+	percentageFull, err := strconv.Atoi(output[:len(output)-1])
+	if err != nil {
+		glog.Warningf("error getting percentage of /var that is free: %v", err)
+	}
+	if percentageFull >= 99 {
+		exit.WithError("", fmt.Errorf("docker daemon out of memory. No space left on device"))
+	}
+
+	if percentageFull > 80 {
```
Comment (medyagh) [inline pkg/minikube/machine/start.go:228]: lets add a TODO for better advice

Comment (priyawadhwa) [reply pkg/minikube/machine/start.go:229]: done

**Pull request merged**